[BEAM-11265] Update quickstart-java.md

pcoet opened the pull request:
I don't think we actually want to use the pom.xml as the `inputFile` for the direct runner. You can use the Shakespeare sample, just like with the other runners.

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.


GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

aaltay commented:
@pcoet - Is there a reason we were using pom.xml? Does gcs path require installing another package or authentication?

TheNeuralBit commented:
I don't *think* authentication is an issue since apache-beam-samples is open publicly, but I think the reason we don't use gs:// in all the examples is it requires `extensions:google-cloud-platform`. This isn't a problem for DataflowRunner though.

Its unfortunate, since running word count on pom.xml is a pretty confusing intro to beam

TheNeuralBit commented:
We could consider making https://github.com/apache/beam/tree/master/sdks/java/maven-archetypes/examples include a local sample file to use with the direct runner.

pcoet commented:
@aaltay and @TheNeuralBit Thanks for the feedback. I'll add one of the example files from the gs bucket to the **examples/**   directory and update the commit.

aaltay commented:
> We could consider making https://github.com/apache/beam/tree/master/sdks/java/maven-archetypes/examples include a local sample file to use with the direct runner.

This is a good idea.

We can also do it similar to python and say `--input /path/to/inputfile`. I do not know if that will increase the barrier for trying.

pcoet commented:
@aaltay and @TheNeuralBit I added a sample.txt with some of the Shakespeare text to https://github.com/apache/beam/tree/master/sdks/java/maven-archetypes/examples 

In the Java getting started guide, I changed each instance of `--inputFile=pom.xml` to `--inputFile=sample.txt`.

aaltay commented:
> @aaltay and @TheNeuralBit I added a sample.txt with some of the Shakespeare text to https://github.com/apache/beam/tree/master/sdks/java/maven-archetypes/examples
> 
> In the Java getting started guide, I changed each instance of `--inputFile=pom.xml` to `--inputFile=sample.txt`.

Sounds reasonable to me. I believe you need to add a license and RAT is failing because of that. (What does it really mean for us to add a license to shakespeare text :) )

pcoet commented:
Shakespeare is licensed! I'm not sure why "Run Java PreCommit" is failing...

aaltay commented:
Java precommit - I think this is a flaky test : "org.apache.beam.sdk.metrics.MetricsTest$AttemptedMetricTests.testAllAttemptedMetrics". @tysonjh @kennknowles - could we deflake this test? Or at least file a jira for it? (link: https://ci-beam.apache.org/job/beam_PreCommit_Java_Commit/14670/testReport/junit/org.apache.beam.sdk.metrics/MetricsTest$AttemptedMetricTests/testAllAttemptedMetrics/)

aaltay commented:
Run Java PreCommit

TheNeuralBit commented:
Looks like the flake is already filed as BEAM-7827

TheNeuralBit commented on website/www/site/content/en/get-started/quickstart-java.md:
```diff
@@ -283,7 +283,7 @@ We are working on adding the instruction for this runner!
 {{< highlight class="runner-spark" >}}
 $ gradle clean execute -DmainClass=org.apache.beam.examples.WordCount \
-    -Dexec.args="--inputFile=pom.xml --output=counts" -Pspark-runner
+    -Dexec.args="--inputFile=sample.txt --output=counts" -Pspark-runner
```
I think we should hold off updating the website until a beam release has gone out with the change to add the sample file. The website is published regularly, so if we update it now the quickstart will start telling users to do this with 2.25 and it won't work.

pcoet replied:
Ah, I didn't know that. Okay, I updated `--inputFile` to follow the Python model, so it points to `/path/to/inputfile`. I still think that makes a lot more sense than using the pom as input.

TheNeuralBit commented on sdks/java/maven-archetypes/examples/sample.txt:
```diff
@@ -0,0 +1,117 @@
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
```
Is it correct to put a license header on this? I think this might qualify for excluding from the RAT checker (I'm assuming we have an allow list somewhere)

aaltay replied:
Yes, there is an allowlist. Here: https://github.com/apache/beam/blob/master/build.gradle#L37

pcoet replied:
I originally left the license off on purpose. It seems like other Apache projects mostly leave the license off of example data, e.g. [apache/spark/blob/master/examples/src/main/resources/kv1.txt](https://github.com/apache/spark/blob/master/examples/src/main/resources/kv1.txt). How would I add the sample to the allow list?

aaltay replied:
If we are switching to use --inputFile=/path/to/inputfile. I guess we can remove the sample completely.

If you woud like to add an exception you can add a line (https://github.com/apache/beam/blob/master/build.gradle#L42) pointing to the sample file.

pcoet commented:
Okay, I think we should be good to go. I left the sample in, but removed the header and excluded it. After the next Beam release, I'll update the docs so that the inputFile points as the included sample. Until then, we can point to /path/to/inputfile.

TheNeuralBit commented on website/www/site/content/en/get-started/quickstart-java.md:
```diff
@@ -266,7 +266,7 @@ For Unix shells (Instructions currently only available for Direct, Spark, and Da
 {{< highlight class="runner-direct">}}
 $ gradle clean execute -DmainClass=org.apache.beam.examples.WordCount \
-    -Dexec.args="--inputFile=pom.xml --output=counts" -Pdirect-runner
+    -Dexec.args="--inputFile=gs://apache-beam-samples/shakespeare/* --output=counts" -Pdirect-runner
```
This one should be /path/to/inputfile as well

```suggestion
    -Dexec.args="--inputFile=/path/to/inputfile --output=counts" -Pdirect-runner
```

TheNeuralBit reviewed (approved):
LGTM, thanks!

I'll commit the change to the direct runner path

TheNeuralBit commented:
Unfortunately since we're just after the 2.26.0 release cut we won't be able to make the website change until 2.27.0 is out (probably early 2021). Filed BEAM-11265 to keep track of this.

TheNeuralBit commented:
Failure looks like flake BEAM-4136

tysonjh commented:
> > Java precommit - I think this is a flaky test : "org.apache.beam.sdk.metrics.MetricsTest$AttemptedMetricTests.testAllAttemptedMetrics". @tysonjh @kennknowles - could we deflake this test? Or at least file a jira for it? (link: https://ci-beam.apache.org/job/beam_PreCommit_Java_Commit/14670/testReport/junit/org.apache.beam.sdk.metrics/MetricsTest$AttemptedMetricTests/testAllAttemptedMetrics/)
> 
> There is an open issue for this,
> 
> https://issues.apache.org/jira/browse/BEAM-1169

Actually, that isn't true. It seems like `distributionMinMax` was introduced to relax the conditions already [1] but this issue was not closed (I'm going to close it).

The flaky bug is https://issues.apache.org/jira/browse/BEAM-7827

kennknowles commented:
A couple years ago I had this same thought ("why are we using pom.xml") and ended up finding an answer to my satisfaction and not changing it... I don't remember why, but please do watch for any issues that might come up. There are no accounts or anything needed for this, right?

TheNeuralBit commented:
> There are no accounts or anything needed for this, right?

See my comment above, I think it's because we didn't want the quickstart project to require the GCP extension:

> I don't think authentication is an issue since apache-beam-samples is open publicly, but I think the reason we don't use gs:// in all the examples is it requires extensions:google-cloud-platform. This isn't a problem for DataflowRunner though.

Ultimately we adjusted this PR, now it just:
- Updates the quickstart archetype by adding a sample.txt file with some shakespeare
- Changes the file paths on the website to `/path/to/input`

Once 2.27.0 is out we can update the website again to point users to the new sample.txt file that they'll have locally. This will probably just be for the DirectRunner since other runners most likely won't have access to the local filesystem.

kennknowles commented:
Great. That's perfect.
